src/util/cow.h

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************\
+
+Module: Copy on write class
+
+Author: Reuben Thomas, reuben.thomas@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_COW_H
+#define CPROVER_UTIL_COW_H
+
+#include "util/small_shared_ptr.h"
+#include "util/invariant.h"
+
+#include <limits>
+
+template <typename T>
+class copy_on_writet final
+{
+public:
+  // Note that this *is* the default constructor. An invariant of a
+  // copy-on-write object is that it is never null.
+  template <typename... Ts>
+  explicit copy_on_writet(Ts &&... ts)
+    : t_(make_small_shared_ptr<T>(std::forward<Ts>(ts)...))
+  {
+    INVARIANT(
+      pointee_is_shareable(*t_),
+      "newly-constructed COW pointers must be shareable");
+  }
+
+  copy_on_writet(const copy_on_writet &rhs)
+    : t_(
+        pointee_is_shareable(*rhs.t_) ? rhs.t_
+                                      : make_small_shared_ptr<T>(*rhs.t_))
+  {
+    INVARIANT(
+      pointee_is_shareable(*t_),
+      "newly-constructed COW pointers must be shareable");
+  }
+
+  copy_on_writet &operator=(const copy_on_writet &rhs)
+  {
+    auto copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  copy_on_writet(copy_on_writet &&rhs)
+  {
+    swap(rhs);
+  }
+
+  copy_on_writet &operator=(copy_on_writet &&rhs)
+  {
+    swap(rhs);
+    return *this;
+  }
+
+  ~copy_on_writet() = default; // delete if the refcount is 1, like usual
+
+  void swap(copy_on_writet &rhs)
+  {
+    std::swap(t_, rhs.t_);
+  }
+
+  const T &read() const
+  {
+    return *t_;
+  }
+
+  T &write(bool mark_shareable)
+  {
+    if(pointee_use_count(*t_) != 1)
+    {
+      t_ = make_small_shared_ptr<T>(*t_);
+    }
+    INVARIANT(
+      pointee_use_count(*t_) == 1,
+      "mutable references to a COW pointer must be unique");
+    pointee_set_shareable(*t_, mark_shareable);
+    INVARIANT(
+      pointee_is_shareable(*t_) == mark_shareable,
```

Comment:
How is it conceivably possible that this invariant is violated right after a call to `pointee_set_shareable`?